dockerdiscovery.go

```diff
@@ -1,44 +1,109 @@
 package dockerdiscovery
 
 import (
-	"context"
-	"errors"
+	"github.com/coredns/coredns/plugin"
+	dockerapi "github.com/fsouza/go-dockerclient"
 	"fmt"
-	"log"
 	"net"
+	"context"
+	"github.com/miekg/dns"
+	"github.com/coredns/coredns/request"
+	"log"
 	"strings"
+	"errors"
+)
+
+type ContainerInfo struct {
+	container *dockerapi.Container
+	address   net.IP
+	domains   []string // resolved domain
+}
+
+type ContainerInfoMap map[string]*ContainerInfo
+
+type ContainerDomainResolver interface {
+	// return domains
+	resolve(container *dockerapi.Container) ([]string, error)
+}
+
+
+
+type SubDomainHostResolver struct {
+	domain string
+}
+func (resolver SubDomainHostResolver) resolve(container *dockerapi.Container) ([]string, error) {
+	var domains []string
+	domains = append(domains, fmt.Sprintf("%s.%s", container.Config.Hostname, resolver.domain))
+	return domains, nil
+}
 
-	"github.com/coredns/coredns/request"
 
-	"github.com/coredns/coredns/plugin"
-	dockerapi "github.com/fsouza/go-dockerclient"
-	"github.com/miekg/dns"
-)
 
-type containerIPMap struct {
-	containerIDHostNameMap map[string][]string
-	byHostName             map[string]net.IP
+type NetworkAliasesResolver struct {
+	network string
 }
+func (resolver NetworkAliasesResolver) resolve(container *dockerapi.Container) ([]string, error) {
+	var domains []string
+
+	if resolver.network != "" {
+		network, ok := container.NetworkSettings.Networks[resolver.network]
+		if ok {
+			domains = append(domains, network.Aliases...)
+		}
+	} else {
+		for _, network := range container.NetworkSettings.Networks {
+			domains = append(domains, network.Aliases...)
+		}
+	}
+
+	for i, d := range domains {
+		domains[i] = fmt.Sprintf("%s.", d)
+	}
+
+	return domains, nil
+}
+
 
 // DockerDiscovery is a plugin that conforms to the coredns plugin interface
 type DockerDiscovery struct {
 	Next           plugin.Handler
 	dockerEndpoint string
-	dockerDomain   string
+	resolvers 	   []ContainerDomainResolver
 	dockerClient   *dockerapi.Client
-	containerIPMap *containerIPMap
+	containerInfoMap   ContainerInfoMap
+	domainIPMap 	map[string]*net.IP
 }
-
 // NewDockerDiscovery constructs a new DockerDiscovery object
-func NewDockerDiscovery(dockerEndpoint string, dockerDomain string) DockerDiscovery {
+func NewDockerDiscovery(dockerEndpoint string) DockerDiscovery {
 	return DockerDiscovery{
 		dockerEndpoint: dockerEndpoint,
-		dockerDomain:   dockerDomain,
-		containerIPMap: &containerIPMap{
-			make(map[string][]string),
-			make(map[string]net.IP),
-		},
+		containerInfoMap:   make(ContainerInfoMap),
+	}
+}
+
+func (dd DockerDiscovery) resolveDomainsByContainer(container *dockerapi.Container) ([]string, error) {
+	var domains []string
+	for _, resolver := range dd.resolvers {
+		var d, _ = resolver.resolve(container)
```

Comment:
Add code to deal with the returned error.  I know `NetworkAliasesResolver` does not return any error for now but may very well be in the future :)